src/mpi/romio/mpi-io/mpioimpl.h

```diff
@@ -18,7 +18,7 @@
 
 #define ROMIO_THREAD_CS_ENTER() MPIR_Ext_cs_enter()
 #define ROMIO_THREAD_CS_EXIT() MPIR_Ext_cs_exit()
-#define ROMIO_THREAD_CS_YIELD() MPIR_Ext_cs_yield()
+#define ROMIO_THREAD_CS_YIELD() MPL_thread_yield()
```

Comment:
Need rename this into `ROMIO_THREAD_YIELD()`. The difference is we are not yielding a critical section here since the calling sites are not inside romio critical section. Since ROMIO never uses `ROMIO_THREAD_CS_YIELD`, we could simply delete that and use `ROMIO_THREAD_YIELD` instead.

Reply:
How about leaving ROMIO_THREAD_CS_YIELD as is, and
adding ROMIO_THREAD_YIELD as a new macro?

Reply:
Sounds good. Maybe also a comment on the difference?

Reply:
I just realized that `MPL_thread_yield()` is already called inside of `MPIR_Ext_cs_yield()`
in file `src/glue/romio/glue_romio.c`. Maybe we do not need to create macro
`ROMIO_THREAD_YIELD` after all?
https://github.com/pmodels/mpich/blob/b6e35732ac6834aa785a6efef35e04a336e58d99/src/glue/romio/glue_romio.c#L111-L117

Reply:
You are missing the point 😄 The whole point here is to call `MPL_thread_yield` but not `MPIR_Ext_cs_exit/enter`, because it is not inside a critical section so can't exit and re-enter. The thread yield is needed only for the case when user-threads (vs kernel preemptive threads) are used and it needs explicit thread yield call.

Reply:
I am seeing an if condition `if (MPIR_ThreadInfo.isThreaded)` in function `MPIR_Ext_cs_exit()` and `MPIR_Ext_cs_enter()`. If the main thread is not in critical section, shouldn't this condition fail, thus effectively go ahead call `MPL_thread_yield()`?
If this is not the case, may be we can add another if condition to check CS and skip calls to `MPIR_Ext_cs_exit/enter`?

Reply:
In the single threaded cases, yes, both `cs_enter/exit` is skipped. We probably should skip `MPL_thread_yield` as well. However, it is the multiple thread case we need these macros and there is a difference between `ROMIO_THREAD_CS_YIELD` and `ROMIO_THREAD_YIELD`.

Reply:
In the multi-threading case. If the calling thread is in a critical section, shouldn't it call  `ROMIO_THREAD_CS_YIELD` instead of `ROMIO_THREAD_YIELD`?
May I suggest to add a check  in function `MPIR_Ext_cs_yield()` for whether the calling thread is in CS or not?

Reply:
I have updated mpich to ensure that we always exit critical section before entering ROMIO so it is guaranteed that romio is not in mpich's critical section. `ROMIO_THREAD_CS_ENTER/EXIT` is for entering and exiting ROMIO's own critical section (for protecting ROMIO's global states). And in this case, since the calling site is not inside ROMIO's critical section, there isn't need and shouldn't call `ROMIO_THREAD_CS_YIELD`.

Reply:
Are you saying functions `ADIOI_GEN_irc_wait_fn()` and `ADIOI_GEN_iwc_wait_fn()` are never called inside ROMIO's critical section?
But I am seeing the followings from running command 'git grep ROMIO_THREAD_CS_'.
```
 mpi-io/iread.c:    ROMIO_THREAD_CS_ENTER();
 mpi-io/iread_all.c:    ROMIO_THREAD_CS_ENTER();
 mpi-io/iwrite.c:    ROMIO_THREAD_CS_ENTER();
 mpi-io/iwrite_all.c:    ROMIO_THREAD_CS_ENTER();
```

Reply:
Ahh I see, you are saying `MPIOI_File_iread` will invoke the `MPI_Wait` ? `ADIOI_GEN_irc_wait_fn` is the hook function for a general request and it is called by `MPI_Wait` family (although I admit I need double check whether there is other places that may also call it). In principle, `ROMIO_THREAD_CS_ENTER/EXIT` should be placed only around where ROMIO accesses its own global state, and it should exit critical section before entering `MPI_Wait`, which will ensure those `wait_fn` are always called outside critical section.

Certainly it is also feasible to have `wait_fn` check whether it is inside ROMIO's critical section, but it can be tricky. It is much simpler to use more granular fashion so it never need yield a critical section.